Homework

desile opened the pull request:
Ларькин П.В.

Commit:
gitignore update for ignoring idea files

Commit:
drafts of hibernate part of work
added entities, dao, and services

Commit:
added jdbc service
added transactionds to jdbc service
added little example in main
fixed nullpointer problem in user deposits

Commit:
not needed update

Commit:
added userservice test

Commit:
newlines

Commit:
added one jdbc test
fix currency timestamp equality

sergey-podunov commented on src/main/java/ru/hh/school/deposits/Deposit.java:
```diff
@@ -0,0 +1,130 @@
+    private Double interest;
+
+    @Column(name = "closed")
+    private Boolean closed;
```
кажется что лучше `boolean`, иначе у тебя появляется третье состояние - "неопределено". И, соответственно, `nullable = false`

sergey-podunov commented on src/main/java/ru/hh/school/deposits/Deposit.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public int hashCode() {
+        return id.hashCode();
```
А если `id == null`? NPE при работе с некоторыми коллекциями?

sergey-podunov commented on src/main/java/ru/hh/school/users/User.java:
```diff
@@ -0,0 +1,112 @@
+    private Set<Deposit> deposits;
+
+    @Column(name = "deleted")
+    private Boolean deleted;
```
та же проблема с `deleted = null` что и в Deposit

sergey-podunov commented on src/main/java/ru/hh/school/users/User.java:
```diff
@@ -0,0 +1,112 @@
+
+    @Override
+    public int hashCode() {
+        return id.hashCode();
```
В Currency hashCode правильно написан, а тут и в Deposit - как подменили :)

desile replied:
Это я как-то невнимательно идеей нагенерил - не сразу обратил внимание, что там для таких случаев можно non-null поля указывать.

sergey-podunov commented on src/main/java/ru/hh/school/deposits/DepositService.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    public Deposit open(User user, double initialAmount, int durationInMonths) {
+        return inTransaction(() -> {
```
У тебя же этот метод не самостоятельный - зачем тут открывается транзакция?

desile replied:
В том виде, в котором он есть сейчас, он потенциально может быть самостоятельным - мы можем передать в него уже созданного пользователя, и для него откроется депозит.

Если вынести операцию над пользователем из DepositService.open в UserService.saveWithDeposit (как предлагается в комментариях ниже) , то тогда уже он перестанет быть самостоятельным, так как его самостоятельный вызов будет приводить к неконсистентному состоянию, так как объект пользователя не будет знать о своем новом депозите.
Тогда и можно убрать транзакцию из этого метода, хотя как я посмотрел inTransaction ничего страшного не делает - если уже есть активная транзакция, то он не будет создавать новую.

sergey-podunov replied:
Если мы передадим снаружи пользователя и внутри проставим ему депозит - значит, у нас пользователь detached entity. И то что внутри мы ему назначаем депозит - не сохранится в базе. Могу что-то путать, но кажется так. И тогда снаружи этого метода все равно придется делать какие-то действия чтоб сохранить информацию о новом депозите (точнее, о связи депозита и пользователя). Т.е. метод получается несамостоятельный, но неявно.

sergey-podunov commented on src/main/java/ru/hh/school/users/UserService.java:
```diff
@@ -0,0 +1,72 @@
+            if(user.id() == null){
+                save(user);
+            }
+            depositService.open(user, initialAmount, durationInMonths);
```
Если читать код отсюда - то кажется что с user ничего не происходит. Однако внутри ему назначается депозит, и об этом не узнаешь пока не зайдешь внутрь метода. Мне кажется более явным все операции с пользователем проводить тут, в `UserService`. Тем более что ты возвращаешь из `DepositService.open()` получившийся депозит.

sergey-podunov commented on src/test/java/ru/hh/school/users/UserServiceTest.java:
```diff
@@ -0,0 +1,145 @@
+        User user1 = new User("first name 1", "last name 1");
+        userService.saveWithDeposit(user1, 12,2000);
+
+        assertNotNull(user1.id());
```
Можно проверить что депозит создался

desile replied:
Создал для этого тест  saveWithDepositShouldOpenDepositForUser()

sergey-podunov commented on src/test/resources/create-tables.sql:
```diff
@@ -0,0 +1,35 @@
+  interest double precision default 0,
+  closed boolean default false
+)
+;
```
В конце обычно перевод строки ставят, чтоб cat удобно использовать

sergey-podunov reviewed (changes requested):
В README.md нет описания предметной области - хотелось бы понимать что происходит до чтения кода. С hashcode проблема, и вопросы про транзакции и неявное изменение user в DepositService

Commit:
added description in readme
made all fields not nullable
fix hashcode function in entities
replaced user operations from deposit service

desile committed:
Update User.java

forgotten boolean
